lib/errors/errors.go

```diff
@@ -0,0 +1,100 @@
+package errors
+
+import (
+	"fmt"
+)
+
+// MultiError is a container for groups of errors.
+type MultiError interface {
+	error
+	// Errors returns all errors carried by this MultiError, or an empty slice otherwise.
+	Errors() []error
+}
+
+// multiError is our default underlying implementation for MultiError. It is compatible
+// with cockroachdb.Error's formatting, printing, etc. and supports introspecting via
+// As, Is, and friends.
+//
+// Implementation is based on https://github.com/knz/shakespeare/blob/master/pkg/cmd/errors.go
+type multiError struct {
+	errs []error
+}
+
+var _ error = (*multiError)(nil)
+var _ MultiError = (*multiError)(nil)
+
+func combineNonNilErrors(err1 error, err2 error) MultiError {
+	var multi1, multi2 MultiError
+	ok1 := As(err1, &multi1)
+	ok2 := As(err2, &multi2)
+	// flatten
+	var errs []error
+	if ok1 && ok2 {
+		errs = append(multi1.Errors(), multi2.Errors()...)
+	} else if ok1 {
+		errs = append(multi1.Errors(), err2)
+	} else if ok2 {
+		errs = append([]error{err1}, multi2.Errors()...)
+	} else {
+		errs = []error{err1, err2}
+	}
+	if len(errs) == 0 {
+		return nil
+	}
+	return &multiError{errs: errs}
+}
+
+// CombineErrors returns a MultiError from err1 and err2. If both are nil, nil is returned.
+func CombineErrors(err1, err2 error) MultiError {
+	if err1 == nil && err2 == nil {
+		return nil
+	}
+	var multi MultiError
+	if err1 == nil {
+		if As(err2, &multi) {
+			return multi
+		}
+		return &multiError{errs: []error{err2}}
+	} else if err2 == nil {
+		if As(err1, &multi) {
+			return multi
+		}
+		return &multiError{errs: []error{err1}}
+	}
+	return combineNonNilErrors(err1, err2)
+}
+
+// Append returns a MultiError created from all given errors, skipping errs that are nil.
+// If no non-nil errors are provided, nil is returned.
+func Append(err error, errs ...error) MultiError {
+	multi := CombineErrors(err, nil)
+	for _, e := range errs {
+		if e != nil {
+			multi = CombineErrors(multi, e)
+		}
+	}
+	return multi
+}
+
+func (e *multiError) Error() string { return fmt.Sprintf("%v", e) }
+func (e *multiError) Errors() []error {
+	if e == nil || e.errs == nil {
+		return []error{}
+	}
+	return e.errs
+}
+
+func (e *multiError) Cause() error  { return e.errs[len(e.errs)-1] }
+func (e *multiError) Unwrap() error { return e.errs[len(e.errs)-1] }
```

Comment:
This doesn't look right. I would expect `Unwrap()` to eventually hit every error in the multierror. This only returns the last error in the slice. Otherwise, `errors.Is(err1, Append(err1, err2))` will return `false`, right? Something like this:

```suggestion
func (e *multiError) Unwrap() error { return &multiError{errs: e.errs[1:] }
```

Reply:
Actually, this doesn't really do it either. However, I think `Unwrap()` is only really used by stdlib when trying to do `errors.Is` and `errors.As`, so if we provide a custom `As` (in addition to the custom `Is` below), I think this would fix my concern.

Reply:
I think you are right, thought curiously without this change the following still works:

```
		{
			name:        "Append",
			err:         Append(errFoo, errBar),
			wantStrings: []string{"foo", "bar"},
			wantIs:      []error{errFoo, errBar},
		},
```

Reply:
I think this is because we provide a custom `Is` below, so calling `errors.Is()` doesn't use `Unwrap()` to get the underlying error. However, I expect `errors.As` doesn't work since we don't (yet) provide a custom implementation.

Reply:
A reproducing test for ya:
```go
type testError1 struct{}

func (t *testError1) Error() string { return "foo" }

type testError2 struct{}

func (t *testError2) Error() string { return "bar" }

func TestAs(t *testing.T) {
	err1 := &testError1{}
	err2 := &testError2{}
	multi := Append(err1, err2)
	var e1 *testError1
	if !As(multi, &e1) {
		t.Fatalf("expected as to work")
	}
}
```

Reply:
^^ note that the above test _does_ work for `testError2` because we unwrap to the last error in the slice

Reply:
> A reproducing test for ya:

Amending d28374e15826c4d2465aa98effaf0dc73b5f1c95 to look more like that I can reproduce the issue 🤔

Reply:
Do the tests you added actually exercise `As`? I don't see any calls to `As` anywhere

Reply:
> Do the tests you added actually exercise As? I don't see any calls to As anywhere

Yeah my bad I did some funky undo - I think it was actually failing, https://github.com/sourcegraph/sourcegraph/commit/d44f7dea994a68db28965ad60841b9a93c4b7091 is the updated test

Reply:
Also, I agree that my original suggestion is wrong. It assumed that we didn't implement a custom `Is` (and even then it wouldn't have worked), but we do. I think all that's needed for my test case to work as expected is a custom `As` method on `multiError` similar to our custom `Is` method.

Reply:
We're close-ish, but the `Append(Append(Append(Append))` test case fails, with the second append failing to assert errBaz:

```
        /Users/robertlin/Projects/sourcegraph/sourcegraph/lib/errors/errors_test.go:87: 4 errors occurred:
            (1) 4 errors occurred:
              | 	* foo
              | 	* bar
              | 	* baz
              | 	* zoo
              | -- details follow
              | (1) foo
              | (1) attached stack trace
              |   -- stack trace:
              |   | github.com/sourcegraph/sourcegraph/lib/errors.TestMultiError
              |   | 	/Users/robertlin/Projects/sourcegraph/sourcegraph/lib/errors/errors_test.go:22
              |   | testing.tRunner
              |   | 	/usr/local/Cellar/go/1.17.2/libexec/src/testing/testing.go:1259
              |   | runtime.goexit
              |   | 	/usr/local/Cellar/go/1.17.2/libexec/src/runtime/asm_amd64.s:1581
              | Wraps: (2) foo
              | Error types: (1) *withstack.withStack (2) *errutil.leafError
              | (2) bar
              | (1) attached stack trace
              |   -- stack trace:
              |   | github.com/sourcegraph/sourcegraph/lib/errors.TestMultiError
              |   | 	/Users/robertlin/Projects/sourcegraph/sourcegraph/lib/errors/errors_test.go:23
              |   | testing.tRunner
              |   | 	/usr/local/Cellar/go/1.17.2/libexec/src/testing/testing.go:1259
              |   | runtime.goexit
              |   | 	/usr/local/Cellar/go/1.17.2/libexec/src/runtime/asm_amd64.s:1581
              | Wraps: (2) bar
              | Error types: (1) *withstack.withStack (2) *errutil.leafError
              | (3) baz
              | (1) baz
              | Error types: (1) *errors.errBazType
              | (4) zoo
              | (1) zoo
              | Error types: (1) *errors.errZooType
            Wraps: (2) zoo
            Error types: (1) *errors.multiError (2) *errors.errZooType
        /Users/robertlin/Projects/sourcegraph/sourcegraph/lib/errors/errors_test.go:98: 
            	Error Trace:	errors_test.go:98
            	Error:      	Should be in error chain:
            	            	expected: %!q(**errors.errBazType=0xc00011a0c8)
            	            	in chain: "4 errors occurred:\n\t* foo\n\t* bar\n\t* baz\n\t* zoo"
            	            		"zoo"
            	Test:       	TestMultiError/Append(Append(Append(Append)))
```